File: advent-of-code/2020/Day9.kt
const val PREAMBLE_SIZE = 25

fun <T> T?.expect(): T = this ?: throw IllegalArgumentException("failed to find weakness")

fun main() {
    val numbers = mapLines { it.toLong() }

    val first = numbers.withIndex()
        .drop(PREAMBLE_SIZE)
        .find { (i, number) ->
            val range = numbers.slice(i - PREAMBLE_SIZE until i)
            range.all { j -> range.all { k -> j + k != number } }
        }
        .expect()
        .value

    var i = 0
    var j = 0
    var sum = 0L

    while (sum != first && j != numbers.size) {
        while (sum < first) sum += numbers[j++]
        if (sum > first) sum -= numbers[i++]
    }

    val range = numbers.slice(i until j)
    val second = range.maxOrNull().expect() + range.minOrNull().expect()

    println("$first $second")
}

File: advent-of-code/2020/Day9.rs
use std::io::prelude::*;
use std::io::stdin;

fn main() {
    const PREAMBLE_SIZE: usize = 25;
    const FAILURE: &str = "failed to find weakness";

    let stdin = stdin();
    let numbers: Vec<_> = stdin.lock()
        .lines()
        .map(|line| line.expect("failed to read from stdin").parse::<i64>().unwrap())
        .collect();

    let first = (PREAMBLE_SIZE..numbers.len() - 1)
        .find(|&i| {
            let range = numbers.get(i - PREAMBLE_SIZE..i).expect(FAILURE);
            return range.iter().all(|&j| range.iter().all(|&k| j + k != numbers[i]));
        })
        .map(|i| numbers[i])
        .expect(FAILURE);

    let mut i = 0;
    let mut j = 0;
    let mut sum: i64 = 0;

    while sum != first && j != numbers.len() {
        while sum < first {
            sum += numbers[j];
            j += 1;
        }
        if sum > first {
            sum -= numbers[i];
            i += 1;
        }
    }

    let range = numbers.get(i..j - 1).expect(FAILURE);
    let second = range.iter().max().expect(FAILURE) + range.iter().min().expect(FAILURE);

    println!("{} {}", first, second);
}
